File: Arrays/Problems/4/solution.py
# In place sorting function using three pointers.
# Time complexity - O(n)
# Space complexity - O(1)

def sort(arr):
    l = 0
    m = 0
    h = len(arr)-1
    while (m<=h):
        if arr[m]==0:
            arr[m], arr[l] = arr[l], arr[m]
            m+=1
            l+=1
        elif arr[m]==1:
            m+=1
        else:
            arr[m], arr[h] = arr[h], arr[m]
            h-=1
    return arr
    

if __name__ == "__main__":
    arr = [1, 0, 2, 1, 1, 2, 0, 0]
    arrSorted = sort(arr)
    print("sorted array: ", arrSorted)
    

File: Arrays/Problems/4/solution.go
package main

import "fmt"

func swap(arr *[]int, i, j int) {
	temp := (*arr)[i]
	(*arr)[i] = (*arr)[j]
	(*arr)[j] = temp
}

// In place sorting function using three pointers.
// Time complexity - O(n)
// Space complexity - O(1)
func sortArr(arr *[]int) {
	l := 0
	m := 0
	h := len(*arr) - 1
	for m <= h {
		if (*arr)[m] == 0 {
			swap(arr, l, m)
			m += 1
			l += 1
		} else if (*arr)[m] == 1 {
			m += 1
		} else {
			swap(arr, m, h)
			h -= 1
		}
	}
}

func main() {
	arr := []int{1, 0, 2, 1, 1, 2, 0, 0}
	sortArr(&arr)
	fmt.Println("sorted array: ", arr)
}
